packages/contracts/src/Inventory.sol

```diff
@@ -246,14 +266,21 @@ contract Inventory is Ownable {
 
     // Checks that the player has all the cards in the given deck in the inventory.
     function checkDeck(address player, uint8 deckID) external view exists(player, deckID) {
-        Deck storage deck = decks[player][deckID];
-        for (uint256 i = 0; i < deck.cards.length ; ++i) {
+        Deck memory deck = decks[player][deckID];
+        uint256 deckLength = deck.cards.length;
+        for (uint256 i = 0; i < deckLength; ++i) {
             uint256 cardID = deck.cards[i];
             if (inventoryCardsCollection.ownerOf(cardID) != player)
                 revert CardNotInInventory(cardID);
+                
+            // check that each card does not exceed its maximum amount of copies
+            uint256 cardCopies;
+            for (uint256 j = 0; j < deckLength; ++j) {
+                if (cardID == deck.cards[j]) ++cardCopies;
+            }
```

Comment:
This is not ideal, as you'll visit the whole deck for each card. So if the deck has size 64 (the max), you'll have 4096 iterations of this, not ideal.

It's also not correct I believe. Say that the deck contains 3 cards with id 1: `[1, 1, 1]` — the inner loop will increment `cardCopies` 3 times on the first iteration, and 3 times on each subsequent iteration, so `cardCopies`will end up with value 9, and not 3 as expected.

To be fair, other solutions suck as well: there is no built-in memory maps, and using a storage map would be too costly in gas (it would actually definitely be ok on Optimism today, and mayyyybe ok too on something like Gnosis chain, but it's not a future-proof solution).

Building a memory based map is possible, but it's a lot of trickiness just for this.

Another similarly fiddly approach: sort the array, then the same cards end up next to each other.
I think that might be the best approach tough.

Here is [some code](https://gist.github.com/subhodi/b3b86cc13ad2636420963e692a4d896f) for sorting an array in Solidity. Sorting is O(n log(n)) on average so for 64 cards, about 384 "iterations". I would benchmark it against [this implementation](https://gist.github.com/sdelvalle57/f5f65a31150ea9321f081630b416ed99) of bubble sort as well, as I'm not sure how much overhead the function call adds in the quicksort version.